app/controllers/homepage_controller.rb

```diff
@@ -5,6 +5,17 @@ class HomepageController < ApplicationController
   def index
     @newbooking = Newbooking.all
     @totalbookings = @newbooking.size
-    @totalmoney = @newbooking.reduce(0) { |sum, booking| sum + booking.service.price }
+    @paidcomplete = calculate_total(Newbooking.where(paid: true, complete: true))
+    @paidincomplete = calculate_total(Newbooking.where(paid: true, complete: false))
+    @unpaidcomplete = calculate_total(Newbooking.where(paid: false, complete: true))
+    @unpaidincomplete = calculate_total(Newbooking.where(paid: false, complete: false))
+  end
+
+  private
+
+  def calculate_total(array)
```

Comment:
great DRYing!

Unfortunately, it introduces a new (minor) issue.

This code now needs to make 5 DB requests, where 4 are subsets of the first. That means that (potentially) we have a lot of data that we have to transfer at least twice, and also have the overhead of making an extra 4 network round-trips to the DB.

Originally, you wrote this calculation of `@totalmoney` slightly differently. Originally, you had a loop that passed through `@newbooking` and calculated the value for `@totalmoney` and `@totalbooking` inside the loop.

Instead of making 4 DB requests to get the data which you already have pulled in during the first request, you could just use that local version and calculate the different values inside the loop.

Another useful method on `enumerable` objects in ruby is `each_with_object`. That is similar to `reduce`, but instead of providing the result of the last iteration, it provides the same object (by reference) each time, and returns the result.

Consider what the below does:

```ruby
hsh = Hash.new { |h, k| h[k] = 0 }

bookings.each_with_object(hsh) do |booking, h|
  h[:stat1] += 1 if booking.should_be_in_stat1
  h[:stat2] += 1 if booking.should_be_in_stat2
end
```

There are a few things with this:
* (in case you've not seen them much before) Hashes are very handy for semi-structured data. They are similar to a `Dictionary` in `C#`
* The `Hash.new` syntax is really useful. This sets a default value that will be returned when there is no value already set
* The `each_with_object` call returns the `hsh` object with all of the values set for the keys




Another (unrelated) thing to consider is what those fields actually mean. This is where additional methods on the model may be useful.

We know that when something is `paid` and `complete`, that means that the booking has been `settled`. When it is is `complete` and not `paid` we know it is `due`. Best practice is to encapsulate the logic in the best place, which for things like this (how two attributes combine to be one concept) is often the model. We would typically "fix" this by adding a method like:
```ruby
def payment_due?
  complete? & !paid?
end
```
to the model, which we would then mean that the controller (or anything else) wouldn't need to understand how we figure out if it is settled or what settled means - it just asks the model.

We would likely have several controllers or other objects that may use the model and need to understand so we can DRY. This is useful because the logic may occasionally change. Suppose next week we change our mind and say "we give our customers 30 days to pay"... Then we would need to also take into account the date to figure out if payment is due. If this logic is in one place (in the model) we only have to change it there.

It also makes our code simpler to test... but we'll get to testing another day ;)